ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/TransactionExtension.java

```diff
@@ -65,27 +63,39 @@
+        
+        bbd.addScope(TransactionScoped.class, true, true);
 
-    public void afterBeanDiscovery(@Observes AfterBeanDiscovery event, BeanManager manager) {        
+        bbd.addAnnotatedType(bm.createAnnotatedType(TransactionalInterceptorMandatory.class), TransactionalInterceptorMandatory.class.getName() + TX_INTERCEPTOR);
+        bbd.addAnnotatedType(bm.createAnnotatedType(TransactionalInterceptorNever.class), TransactionalInterceptorNever.class.getName() + TX_INTERCEPTOR);
+        bbd.addAnnotatedType(bm.createAnnotatedType(TransactionalInterceptorNotSupported.class), TransactionalInterceptorNotSupported.class.getName() + TX_INTERCEPTOR);
+        bbd.addAnnotatedType(bm.createAnnotatedType(TransactionalInterceptorRequired.class), TransactionalInterceptorRequired.class.getName() + TX_INTERCEPTOR);
+        bbd.addAnnotatedType(bm.createAnnotatedType(TransactionalInterceptorRequiresNew.class), TransactionalInterceptorRequiresNew.class.getName() + TX_INTERCEPTOR);
+        bbd.addAnnotatedType(bm.createAnnotatedType(TransactionalInterceptorSupports.class), TransactionalInterceptorSupports.class.getName() + TX_INTERCEPTOR);
+    }
 
-        boolean maybeAddInitialContextBean = false;
-        
-        Set<Bean<?>> beans = manager.getBeans(TransactionManager.class);
-        if (beans.isEmpty()) {
-            event.addBean(new JNDIBean<>(jtaPropertyManager.getJTAEnvironmentBean().getTransactionManagerJNDIContext(), TransactionManager.class));
-            maybeAddInitialContextBean = true;
-        }
+    /**
+     * Gathering information about managed bean to obtain mapping bean to annotated type.
+     * This is needed later when handling Stereotypes in TransactionalInterceptorBase.
+     */
+    public void processManagedBean(@Observes ProcessManagedBean<?> pmb) {
+        beanToAnnotatedTypeMapping.put(pmb.getBean(), pmb.getAnnotatedBeanClass());
+    }
 
-        beans = manager.getBeans(TransactionSynchronizationRegistry.class);
-        if (beans.isEmpty()) {
-            event.addBean(new JNDIBean<>(jtaPropertyManager.getJTAEnvironmentBean().getTransactionSynchronizationRegistryJNDIContext(), TransactionSynchronizationRegistry.class));
-            maybeAddInitialContextBean = true;
-        }
+    public void afterBeanDiscovery(@Observes AfterBeanDiscovery event, BeanManager manager) {        
 
-        if (maybeAddInitialContextBean) {
+        Set<Bean<?>> beans = manager.getBeans(TransactionSynchronizationRegistry.class);
+        if (beans == null || beans.isEmpty()) {
+            event.addBean(new JNDIBean<>(jtaPropertyManager.getJTAEnvironmentBean().getTransactionSynchronizationRegistryJNDIContext(), TransactionSynchronizationRegistry.class));
             beans = manager.getBeans(InitialContext.class);
-            if (beans.isEmpty()) {
+            if (beans == null || beans.isEmpty()) {
                 event.addBean(new AbstractBean<InitialContext>() {
```

Comment:
@ljnelson sorry for asking a little bit unrelated thing to this PR but when I'm going through the code I'm thinking that this part where there is created `new AbstractBean<InitialContext>` is a bit hard to read. Do you think it's reasonable to pull the definition off and creating a new class that would be just referenced here? Similar to how `JNDIBean` is defined?

Reply:
I can do this, sure.  Given the discussion going on on [WFLY-11858](https://issues.jboss.org/browse/WFLY-11858), would you like me to make its scope `Dependent` as well?

Reply:
@ljnelson as I commented on the WFLY-11858 I think the `Singleton` scope sound fine to me.

Reply:
I've resolved this in my latest pushes by incorporating feedback from [WFLY-11858](https://issues.jboss.org/browse/WFLY-11858)—I've removed all the acquisition of `InitialContext` instances by any means other than direct construction.  This issue is therefore resolved.